src/exp.ml

```diff
@@ -949,8 +952,12 @@ and of_let
       | Texp_function _ -> false
       | _ -> true
       end ->
-      Type.of_typ_expr true typ_vars exp_type >>= fun (_, _, new_typ_vars) ->
-      return (List.length new_typ_vars <> 0)
```

Comment:
With the new algorithm `new_typs_vars == 0` does not guarantee that this is not a function, we need to check how `typ_vars` have changed for that.